### PR TITLE
Improving the multiparty key-switching API and cross-scheme implementation

### PR DESCRIPTION
**Pull request by ChristianMct**
This PR continues the effort of #100,  by providing common interfaces and implementation base for the multiparty key-switching protocols (and derivatives) between the `dbfv` and `dckks` packages.

- [x] Centralize key-establishement and their marshalling of DBFV and DCKKS in DRLWE. Will require RLWE to have default parameters for testing and centralized rlwe.keygen and key.switching.

### CODE REVIEW

File: dbfv/shares_to_enc.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	ringQ   *ring.Ring
+	encoder bfv.Encoder
+
+	zero         *rlwe.SecretKey
+	tmpPlaintext *bfv.Plaintext
+}
+
+// NewS2EProtocol creates a new S2EProtocol struct from the passed BFV parameters.
+func NewS2EProtocol(params bfv.Parameters, sigmaSmudging float64) *S2EProtocol {
+	e2s := new(S2EProtocol)
```
[the remark this answers is not in the document]
Comment (ChristianMct) [reply dbfv/shares_to_enc.go]: Fixed

File: dbfv/shares_to_enc.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+// polynomial sampled from the CRS `c1` and the party's secret share of the message.
+func (s2e *S2EProtocol) GenShare(sk *rlwe.SecretKey, c1 *ring.Poly, secretShare rlwe.AdditiveShare, c0ShareOut *drlwe.CKSShare) {
+	s2e.encoder.ScaleUp(&bfv.PlaintextRingT{Plaintext: &rlwe.Plaintext{Value: &secretShare.Value}}, s2e.tmpPlaintext)
+	s2e.CKSProtocol.GenShare(s2e.zero, sk, &rlwe.Element{Value: []*ring.Poly{c0ShareOut.Value, c1}}, c0ShareOut)
+	s2e.ringQ.Add(c0ShareOut.Value, s2e.tmpPlaintext.Value, c0ShareOut.Value)
+}
+
+// GetEncryption computes the final encryption of the secret-shared message when provided with the aggregation `c0Agg` of the parties'
+// share in the protocol and with the common, CRS-sampled polynomial `c1`.
+func (s2e *S2EProtocol) GetEncryption(c0Agg *drlwe.CKSShare, c1 *ring.Poly, ctOut bfv.Ciphertext) {
```
[the remark this answers is not in the document]
Comment (ChristianMct) [reply dbfv/shares_to_enc.go]: Fixed

File: dbfv/shares_to_enc.go [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+	e2s.ringQ = params.RingQ()
+	e2s.encoder = bfv.NewEncoder(params)
+	e2s.zero = rlwe.NewSecretKey(params.Parameters)
+	e2s.tmpPlaintext = bfv.NewPlaintext(params)
+	return e2s
+}
+
+// GenShare generates a party's in the shares-to-encryption protocol given the party's secret-key share `sk`, a common
+// polynomial sampled from the CRS `c1` and the party's secret share of the message.
+func (s2e *S2EProtocol) GenShare(sk *rlwe.SecretKey, c1 *ring.Poly, secretShare rlwe.AdditiveShare, c0ShareOut *drlwe.CKSShare) {
```
[the remark this answers is not in the document]
Comment (ChristianMct) [reply dbfv/shares_to_enc.go]: Fixed

File: dbfv/dbfv_test.go [hunk 4]
```diff
@@ -465,6 +484,75 @@ func testRotKeyGenRotCols(testCtx *testContext, t *testing.T) {
...
+		fmt.Println("Have:", res[:see])
+
+		assert.True(t, utils.EqualSliceUint64(coeffs, res))
+
+	})
+
+	crs := ring.NewUniformSampler(testCtx.prng, testCtx.ringQ)
+	c1 := crs.ReadNew()
+
+	t.Run(testString("S2EProtocol/", parties, testCtx.params), func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ChristianMct) [reply dbfv/dbfv_test.go:542]: Fixed

**Comment by ChristianMct** [comment]
Todos before merging:

- [ ] reorganize files in distributed scheme
- [ ] update changelog

**Review by jrtroncoso — APPROVED** [review]
I've done a quick pass and updated the CHANGELOG

**Pull request merged**